src/python/pants/backend/go/module.py

```diff
@@ -174,6 +179,44 @@ async def resolve_go_module(
     )
 
 
+@dataclass(frozen=True)
+class FindOwningGoModuleRequest:
+    address: Address
+
+
+@dataclass(frozen=True)
+class ResolvedOwningGoModule:
+    module_address: Optional[Address]
+
+
+@rule
+async def find_nearest_go_module(request: FindOwningGoModuleRequest) -> ResolvedOwningGoModule:
```

Comment:
This was refactored out of the `target_type_rules` to allow reuse. Probably can be a separate PR.